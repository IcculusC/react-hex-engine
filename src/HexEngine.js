import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { HexEngineProvider } from "./Context";
import Point from "./models/Point";
import Orientation from "./models/Orientation";

function getPointOffset(corner, orientation, size) {
  let angle = (2.0 * Math.PI * (corner + orientation.startAngle)) / 6;
  return new Point(size.x * Math.cos(angle), size.y * Math.sin(angle));
}

function calculateCoordinates(orientation, size) {
  const center = new Point(0, 0);
  return [...Array(6).keys()].map((_, i) => {
    const offset = getPointOffset(i, orientation, size);
    return new Point(center.x + offset.x, center.y + offset.y);
  });
}

const HexEngine = props => {
  const {
    children,
    classes,
    flat,
    height,
    origin,
    size,
    spacing,
    viewBox,
    width
  } = props;

  const orientation = flat ? Orientation.Flat : Orientation.Pointy;
  const [points, setPoints] = useState(
    calculateCoordinates(orientation, size)
      .map(point => point.toString())
      .join(" ")
  );
  useEffect(
    () => {
      setPoints(() =>
        calculateCoordinates(orientation, size)
          .map(point => point.toString())
          .join(" ")
      );
    },
    [orientation, size]
  );
  const layout = { orientation, origin, size, spacing };

  return (
    <svg
      key={JSON.stringify(layout)}
      className={classNames("grid", classes.grid)}
      height={height}
      version="1.1"
      viewBox={`${viewBox.x} ${viewBox.y} ${viewBox.width} ${viewBox.height}`}
      width={width}
      xmlns="http://www.w3.org/2000/svg"
    >
      <HexEngineProvider
        value={{
          layout,
          points,
          viewBox
        }}
      >
        <g className={classes.layout}>{children}</g>
      </HexEngineProvider>
    </svg>
  );
};

HexEngine.propTypes = {
  children: PropTypes.node.isRequired,
  classes: PropTypes.objectOf(PropTypes.string),
  /** Determines if the hexagons are oriented with a point or an edge facing up */
  flat: PropTypes.bool,
  /** CSS string or number */
  height: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.number.isRequired
  ]),
  /** Origin of the grid */
  origin: PropTypes.oneOfType([
    PropTypes.instanceOf(Point),
    PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired
    })
  ]),
  /** Size of the hexagons in each dimension */
  size: PropTypes.oneOfType([
    PropTypes.instanceOf(Point),
    PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired
    })
  ]),
  /** Space between hexagons */
  spacing: PropTypes.number,
  /** CSS string or number */
  width: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.number.isRequired
  ]),
  /** The viewBox {x,y,width,height} of the svg view area */
  viewBox: PropTypes.shape({
    height: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired
  })
};

HexEngine.defaultProps = {
  classes: { grid: "", layout: "" },
  flat: true,
  height: 480,
  origin: new Point(0, 0),
  size: new Point(10, 10),
  spacing: 1.0,
  width: 640,
  viewBox: {
    height: 100,
    width: 100,
    x: -50,
    y: -50
  }
};

export const HexEngine_ = HexEngine;
export default HexEngine_;
