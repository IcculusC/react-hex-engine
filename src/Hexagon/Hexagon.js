import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Hex from "../models/Hex";
import HexUtils from "../HexUtils";
import Point from "../models/Point";
import { withExpandedLayout } from "../Context";

class Hexagon extends Component {
  static propTypes = {
    children: PropTypes.node,
    classes: PropTypes.objectOf(PropTypes.any),
    data: PropTypes.object,
    highlighted: PropTypes.bool,
    layout: PropTypes.objectOf(PropTypes.any).isRequired,
    onClick: PropTypes.func,
    onDragEnd: PropTypes.func,
    onDragOver: PropTypes.func,
    onDragStart: PropTypes.func,
    onDrop: PropTypes.func,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
    onMouseOver: PropTypes.func,
    points: PropTypes.string.isRequired,
    q: PropTypes.number.isRequired,
    r: PropTypes.number.isRequired,
    s: PropTypes.number.isRequired,
    selected: PropTypes.bool,
    showCoordinates: PropTypes.bool
  };

  static defaultProps = {
    className: "",
    classes: {
      group: "",
      hexagon: "",
      highlighted: "",
      hovered: "",
      polygon: "",
      q: "",
      r: "",
      s: "",
      selected: ""
    },
    showCoordinates: false
  };

  static getCoordinateTextOffset(
    corner,
    { orientation, size },
    offset = new Point(0, 0),
    scale = new Point(0.75, 0.75)
  ) {
    const angle = (2.0 * Math.PI * (corner + 0.5)) / 6;
    return new Point(
      size.x * scale.x * Math.cos(angle),
      size.y * scale.y * Math.sin(angle)
    );
  }

  state = {
    hex: {},
    hovered: false
  };

  constructor(props) {
    super(props);
    const hex = new Hex(props.q, props.r, props.s);
    const pixel = HexUtils.hexToPixel(hex, props.layout);
    this.state = {
      ...this.state,
      hex,
      pixel
    };
  }

  onClick(e) {
    if (this.props.onClick) {
      this.props.onClick(e, this.state.hex);
    }
  }

  onDragEnd(e) {
    if (this.props.onDragEnd) {
      e.preventDefault();
      const success = e.dataTransfer.dropEffect !== "none";
      this.props.onDragEnd(e, this.state.hex, success);
    }
  }

  onDragOver(e) {
    if (this.props.onDragOver) {
      this.props.onDragOver(e, this.state.hex);
    }
  }

  onDragStart(e) {
    if (this.props.onDragStart) {
      const targetProps = {
        ...this.state,
        data: this.props.data,
        classes: this.props.classes
      };
      e.dataTransfer.setData("hexagon", JSON.stringify(targetProps));
      this.props.onDragStart(e, this.state.hex);
    }
  }

  onDrop(e) {
    if (this.props.onDrop) {
      e.preventDefault();
      const target = JSON.parse(e.dataTransfer.getData("hexagon"));
      this.props.onDrop(e, this.state.hex, target);
    }
  }

  onMouseEnter(e) {
    this.setState({ hovered: true });
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(e, this.state.hex);
    }
  }

  onMouseLeave(e) {
    this.setState({ hovered: false });
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(e, this.state.hex);
    }
  }

  onMouseOver(e) {
    if (this.props.onMouseOver) {
      this.props.onMouseOver(e, this.state.hex);
    }
  }

  render() {
    const {
      classes,
      highlighted,
      layout,
      points,
      q,
      r,
      s,
      selected,
      showCoordinates
    } = this.props;
    const { hex, hovered, pixel } = this.state;
    let qPixel, rPixel, sPixel;
    if (showCoordinates) {
      qPixel = Hexagon.getCoordinateTextOffset(3, layout, {
        x: 0,
        y: 1
      });
      rPixel = Hexagon.getCoordinateTextOffset(1, layout, {
        x: -1,
        y: -1
      });
      sPixel = Hexagon.getCoordinateTextOffset(5, layout, {
        x: -2,
        y: 1
      });
    }
    return (
      <g
        className={classNames("hexagon-group", classes.group)}
        draggable="true"
        onClick={e => this.onClick(e)}
        onDragEnd={e => this.onDragEnd(e)}
        onDragOver={e => this.onDragOver(e)}
        onDragStart={e => this.onDragStart(e)}
        onDrop={e => this.onDrop(e)}
        onMouseEnter={e => this.onMouseEnter(e)}
        onMouseLeave={e => this.onMouseLeave(e)}
        onMouseOver={e => this.onMouseOver(e)}
        transform={`translate(${pixel.x}, ${pixel.y})`}
      >
        <g
          className={classNames("hexagon", classes.hexagon, {
            [classes.selected]: selected,
            [classes.highlighted]: highlighted,
            [classes.hovered]: hovered
          })}
        >
          <polygon className={classes.polygon} points={points} />
          {this.props.children}
          {showCoordinates && (
            <React.Fragment>
              <text
                {...qPixel}
                className={classes.q}
                fontSize={2}
                textAnchor="middle"
              >
                {q}
              </text>
              <text
                {...rPixel}
                className={classes.r}
                fontSize={2}
                textAnchor="middle"
              >
                {r}
              </text>
              <text
                {...sPixel}
                className={classes.s}
                fontSize={2}
                textAnchor="middle"
              >
                {s}
              </text>
            </React.Fragment>
          )}
        </g>
      </g>
    );
  }
}

export default withExpandedLayout(Hexagon);
