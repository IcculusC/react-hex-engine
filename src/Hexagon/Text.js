import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

class Text extends Component {
  static propTypes = {
    anchor: PropTypes.string,
    children: PropTypes.string,
    className: PropTypes.string,
    classes: PropTypes.objectOf(PropTypes.string),
    x: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    y: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  };

  static defaultProps = {
    anchor: "middle",
    classes: { text: "" }
  };

  render() {
    const { anchor, children, className, classes, x, y } = this.props;
    return (
      <text
        className={classNames(className, classes.text)}
        textAnchor={anchor}
        x={x || 0}
        y={y ? y : "0.3em"}
      >
        {children}
      </text>
    );
  }
}

export default Text;
