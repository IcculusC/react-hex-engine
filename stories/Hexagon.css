.hexagon {
    fill: #00a;
    transition: fill 0.2s;
}

.hexagon.hovered {
    fill: #aa0;
    transition: fill 0.2s;
}

.hexagon.hovered > polygon {
    stroke: #a00;
    stroke-width: 0.1;
}

.hexagon.selected {
    fill: #ff0;
    transition: fill 0.5s;
}

.hexagon.selected > polygon {
    stroke: #a00;
    stroke-width: 0.1;
}

.hexagon.selected.hovered {
    fill: #fff;
}

.hexagon.selected.hovered > polygon {
    stroke: #a00;
    stroke-width: 0.1;
}

.hexagon.highlighted {
    fill: blue;
}

.hexagon.showCoordinates > text.axis {
    fill: #fff;
    font-size: 2px;
    font-family: sans-serif;
}

.hexagon.text > text.text {
    fill: #fff;
    font-size: 3px;
    font-family: sans-serif;
}
