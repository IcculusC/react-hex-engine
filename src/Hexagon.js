import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Hex from "./models/Hex";
import HexUtils from "./HexUtils";
import Point from "./models/Point";
import Text from "./Text";
import { withHexEngine } from "./Context";

class Hexagon extends Component {
  static propTypes = {
    children: PropTypes.node,
    classes: PropTypes.objectOf(PropTypes.any),
    data: PropTypes.object,
    hoverable: PropTypes.bool,
    /** Provides a means to highlight a tile without selecting it, styleable with the `highlighted` class */
    highlighted: PropTypes.bool,
    layout: PropTypes.objectOf(PropTypes.any).isRequired,
    onClick: PropTypes.func,
    onDragEnd: PropTypes.func,
    onDragOver: PropTypes.func,
    onDragStart: PropTypes.func,
    onDrop: PropTypes.func,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
    onMouseOver: PropTypes.func,
    points: PropTypes.string.isRequired,
    q: PropTypes.number.isRequired,
    r: PropTypes.number.isRequired,
    s: PropTypes.number.isRequired,
    /** Disable the `selected` prop, rather redundant */
    selectable: PropTypes.bool,
    selected: PropTypes.bool,
    /** Show the `q`, `r`, `s` coordinates of the hexagon, styleable with the `q`, `r`, and `s` classes */
    showCoordinates: PropTypes.bool,
    /** Show text in the hexagon, styleable with the `text` class */
    text: PropTypes.string,
    TextProps: PropTypes.objectOf(PropTypes.any)
  };

  static defaultProps = {
    classes: {
      group: "",
      hexagon: "",
      highlighted: "",
      hovered: "",
      polygon: "",
      q: "",
      r: "",
      s: "",
      selected: "",
      text: ""
    },
    data: {},
    highlighted: false,
    hoverable: true,
    selectable: true,
    selected: false,
    showCoordinates: false,
    text: "",
    TextProps: {}
  };

  static getCoordinateTextOffset(
    corner,
    { orientation, size },
    offset = new Point(0, 0),
    scale = new Point(0.75, 0.75)
  ) {
    const angle = (2.0 * Math.PI * (corner + 0.5)) / 6;
    return new Point(
      size.x * scale.x * Math.cos(angle),
      size.y * scale.y * Math.sin(angle)
    );
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.points !== nextProps.points) {
      const { layout, points, q, r } = nextProps;
      const hex = new Hex(q, r);
      const pixel = HexUtils.hexToPixel(hex, nextProps.layout);
      return { hex, pixel, layout, points };
    }
    return null;
  }

  state = {
    hex: {},
    hovered: false,
    layout: {},
    pixel: {},
    points: ""
  };

  constructor(props) {
    super(props);
    const hex = new Hex(props.q, props.r);
    const pixel = HexUtils.hexToPixel(hex, props.layout);
    this.state = {
      ...this.state,
      hex,
      pixel
    };
  }

  onClick(e) {
    if (this.props.onClick) {
      this.props.onClick(e, this.state.hex);
    }
  }

  onDragEnd(e) {
    if (this.props.onDragEnd) {
      e.preventDefault();
      const success = e.dataTransfer.dropEffect !== "none";
      this.props.onDragEnd(e, this.state.hex, success);
    }
  }

  onDragOver(e) {
    if (this.props.onDragOver) {
      this.props.onDragOver(e, this.state.hex);
    }
  }

  onDragStart(e) {
    if (this.props.onDragStart) {
      const targetProps = {
        ...this.state,
        data: this.props.data,
        classes: this.props.classes
      };
      e.dataTransfer.setData("hexagon", JSON.stringify(targetProps));
      this.props.onDragStart(e, this.state.hex);
    }
  }

  onDrop(e) {
    if (this.props.onDrop) {
      e.preventDefault();
      const target = JSON.parse(e.dataTransfer.getData("hexagon"));
      this.props.onDrop(e, this.state.hex, target);
    }
  }

  onMouseEnter(e) {
    this.setState({ hovered: true });
    if (this.props.onMouseEnter) {
      this.props.onMouseEnter(e, this.state.hex);
    }
  }

  onMouseLeave(e) {
    this.setState({ hovered: false });
    if (this.props.onMouseLeave) {
      this.props.onMouseLeave(e, this.state.hex);
    }
  }

  onMouseOver(e) {
    if (this.props.onMouseOver) {
      this.props.onMouseOver(e, this.state.hex);
    }
  }

  render() {
    const {
      classes,
      highlighted,
      hoverable,
      layout,
      points,
      q,
      r,
      s,
      selectable,
      selected,
      showCoordinates,
      text,
      TextProps
    } = this.props;
    const { hovered, pixel } = this.state;
    let qPixel, rPixel, sPixel;
    if (showCoordinates) {
      qPixel = Hexagon.getCoordinateTextOffset(3, layout, {
        x: 0,
        y: 1
      });
      rPixel = Hexagon.getCoordinateTextOffset(1, layout, {
        x: -1,
        y: -1
      });
      sPixel = Hexagon.getCoordinateTextOffset(5, layout, {
        x: -2,
        y: 1
      });
    }
    return (
      <g
        className={classNames("hexagon-group", classes.group)}
        draggable="true"
        onClick={e => this.onClick(e)}
        onDragEnd={e => this.onDragEnd(e)}
        onDragOver={e => this.onDragOver(e)}
        onDragStart={e => this.onDragStart(e)}
        onDrop={e => this.onDrop(e)}
        onMouseEnter={e => this.onMouseEnter(e)}
        onMouseLeave={e => this.onMouseLeave(e)}
        onMouseOver={e => this.onMouseOver(e)}
        transform={`translate(${pixel.x}, ${pixel.y})`}
      >
        <g
          className={classNames("hexagon", classes.hexagon, {
            [classes.selected]: selectable && selected,
            [classes.highlighted]: highlighted,
            [classes.hovered]: hoverable && hovered
          })}
        >
          <polygon className={classes.polygon} points={points} />
          {this.props.children}
          {text ? (
            <Text classes={{ text: classes.text }} {...TextProps}>
              {text}
            </Text>
          ) : null}
          {showCoordinates ? (
            <React.Fragment>
              <text
                {...qPixel}
                className={classes.q}
                fontSize={2}
                textAnchor="middle"
              >
                {q}
              </text>
              <text
                {...rPixel}
                className={classes.r}
                fontSize={2}
                textAnchor="middle"
              >
                {r}
              </text>
              <text
                {...sPixel}
                className={classes.s}
                fontSize={2}
                textAnchor="middle"
              >
                {s}
              </text>
            </React.Fragment>
          ) : null}
        </g>
      </g>
    );
  }
}

export const Hexagon_ = Hexagon;
export default withHexEngine(Hexagon_);
