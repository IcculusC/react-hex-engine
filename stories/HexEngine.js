import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, boolean, number, object } from "@storybook/addon-knobs";
import { withInfo } from "@storybook/addon-info";
import GridGenerator from "../src/GridGenerator";
import Hexagon, { Hexagon_ } from "../src/Hexagon";
import HexEngine from "../src/HexEngine";
import Point from "../src/models/Point";
import IndexedEngine from "../src/IndexedEngine";
import "./Hexagon.css";

Hexagon.displayName = "Hexagon";
Hexagon_.displayName = "Hexagon";
HexEngine.displayName = "HexEngine";

const stories = storiesOf("HexEngine", module);

stories.addDecorator(withKnobs);

stories.addDecorator(
  withInfo({
    header: false,
    inline: true,
    maxPropStringLength: 512,
    source: true,
    propTablesExclude: [Hexagon_, Hexagon]
  })
);

stories
  .add(
    "Kitchen Sink",
    () => {
      const flat = boolean("flat", true, "HexEngine");
      const size = object("size", { x: 10, y: 10 }, "HexEngine");
      const origin = object("origin", { x: 0, y: 0 }, "HexEngine");
      const spacing = number("spacing", 1.0, {}, "HexEngine");
      const width = number("width", 320, {}, "HexEngine");
      const height = number("height", 240, {}, "HexEngine");
      const viewBox = object(
        "viewBox",
        { x: -30, y: -30, width: 60, height: 60 },
        "HexEngine"
      );

      return (
        <HexEngine
          width={width}
          height={height}
          viewBox={viewBox}
          flat={flat}
          size={size}
          origin={origin}
          spacing={spacing}
        >
          {GridGenerator.hexagon(1).map(hex => (
            <Hexagon {...hex} key={hex.toString()} />
          ))}
        </HexEngine>
      );
    },
    {
      info: {
        propTables: [HexEngine]
      }
    }
  )
  .add("non zero origin", () => (
    <HexEngine
      origin={new Point(-5, -5)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("pointy orientation", () => (
    <HexEngine
      flat={false}
      width={320}
      height={240}
      viewBox={{
        x: -25,
        y: -25,
        width: 50,
        height: 50
      }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("bigger size", () => (
    <HexEngine
      size={new Point(30, 30)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("smaller size", () => (
    <HexEngine
      size={new Point(3, 3)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("irregular proportions", () => (
    <HexEngine
      size={new Point(15, 5)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("spacing", () => (
    <HexEngine
      spacing={1.05}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={-1} r={0} s={1} />
      <Hexagon q={0} r={0} s={0} />
      <Hexagon q={1} r={0} s={-1} />
    </HexEngine>
  ))
  .add("indexed", () => {
    const viewBox = object(
      "viewBox",
      { x: -50, y: -50, width: 100, height: 100 },
      "HexEngine"
    );
    const map = GridGenerator.hexagon(50);
    const classes = {
      hexagon: "showCoordinates text",
      highlighted: "highlighted",
      hovered: "hovered",
      selected: "selected",
      q: "axis",
      r: "axis",
      s: "axis",
      text: "text"
    };
    return (
      <IndexedEngine
        height={400}
        map={map}
        viewBox={viewBox}
        width={400}
        spacing={1.05}
      >
        {hexes =>
          hexes.map(hex => (
            <Hexagon
              {...hex}
              classes={classes}
              showCoordinates
              selected={Math.random() > 0.5}
            />
          ))
        }
      </IndexedEngine>
    );
  });
