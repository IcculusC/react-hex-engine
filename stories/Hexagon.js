import React from "react";
import { storiesOf } from "@storybook/react";
import addons from "@storybook/addons";
import Events from "@storybook/core-events";
import { withKnobs, text, boolean } from "@storybook/addon-knobs";
import { withInfo } from "@storybook/addon-info";
import Hexagon, { Hexagon_ } from "../src/Hexagon";
import HexEngine from "../src/HexEngine";
import Point from "../src/models/Point";
import { manager } from "@storybook/addon-knobs/dist/registerKnobs";
import "./Hexagon.css";

Hexagon.displayName = "Hexagon";
Hexagon_.displayName = "Hexagon";
HexEngine.displayName = "HexEngine";

const stories = storiesOf("Hexagon", module);

stories.addDecorator(withKnobs);
stories.addDecorator(withInfo);

stories.add(
  "Kitchen Sink",
  () => {
    const selectable = boolean("selectable", true, "Hexagon");
    const selected = boolean("selected", false, "Hexagon");
    const highlighted = boolean("highlighted", false, "Hexagon");
    const hoverable = boolean("hoverable", true, "Hexagon");
    const showCoordinates = boolean("showCoordinates", false, "Hexagon");
    const innerText = text("text", "", "Hexagon");

    return (
      <HexEngine
        size={new Point(10, 10)}
        width={320}
        height={240}
        viewBox={{
          x: -25,
          y: -25,
          width: 50,
          height: 50
        }}
      >
        <Hexagon
          q={0}
          r={0}
          s={0}
          selectable={selectable}
          selected={selected}
          highlighted={highlighted}
          hoverable={hoverable}
          text={innerText}
          showCoordinates={showCoordinates}
          classes={{
            hexagon: "showCoordinates text",
            highlighted: "highlighted",
            hovered: "hovered",
            selected: "selected",
            q: "axis",
            r: "axis",
            s: "axis",
            text: "text"
          }}
          onClick={() => {
            const { knobStore } = manager;
            const knobOptions = knobStore.get("selected");
            knobOptions.value = !selected;
            knobStore.markAllUnused();
            addons.getChannel().emit(Events.FORCE_RE_RENDER);
          }}
        />
      </HexEngine>
    );
  },
  {
    info: {
      header: false,
      inline: true,
      maxPropStringLength: 512,
      propTables: [Hexagon_], // should be able to fix this with useContext in react 16.7.x
      propTablesExclude: [HexEngine, Hexagon],
      source: false,
      excludedPropTypes: [
        "data",
        "layout",
        "onDragEnd",
        "onDragOver",
        "onDragStart",
        "onDragStop",
        "onDrop",
        "points"
      ]
    }
  }
);
