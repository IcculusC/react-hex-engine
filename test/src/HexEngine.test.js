import React from "react";
import renderer from "react-test-renderer";
import Hexagon from "../../src/Hexagon";
import HexEngine from "../../src/HexEngine";

describe("HexEngine", () => {
  it("HexEngine should render correctly with default props", () => {
    const tree = renderer
      .create(
        <HexEngine classes={{ layout: "test1" }}>
          <Hexagon q={0} r={0} s={0} />
        </HexEngine>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("HexEngine should render correctly with pointy orientation", () => {
    const tree = renderer
      .create(
        <HexEngine classes={{ layout: "test1" }} flat={false}>
          <Hexagon q={0} r={0} s={0} />
        </HexEngine>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
