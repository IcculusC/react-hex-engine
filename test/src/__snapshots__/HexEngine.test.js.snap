// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HexEngine HexEngine should render correctly with default props 1`] = `
<svg
  className="grid"
  height={480}
  version="1.1"
  viewBox="-50 -50 100 100"
  width={640}
  xmlns="http://www.w3.org/2000/svg"
>
  <g
    className="test1"
  >
    <g
      className="hexagon-group"
      draggable="true"
      onClick={[Function]}
      onDragEnd={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      transform="translate(0, 0)"
    >
      <g
        className="hexagon"
      >
        <polygon
          className=""
          points="10,0 5.000000000000001,8.660254037844386 -4.999999999999998,8.660254037844387 -10,1.2246467991473533e-15 -5.000000000000004,-8.660254037844386 5.000000000000001,-8.660254037844386"
        />
      </g>
    </g>
  </g>
</svg>
`;

exports[`HexEngine HexEngine should render correctly with pointy orientation 1`] = `
<svg
  className="grid"
  height={480}
  version="1.1"
  viewBox="-50 -50 100 100"
  width={640}
  xmlns="http://www.w3.org/2000/svg"
>
  <g
    className="test1"
  >
    <g
      className="hexagon-group"
      draggable="true"
      onClick={[Function]}
      onDragEnd={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      transform="translate(0, 0)"
    >
      <g
        className="hexagon"
      >
        <polygon
          className=""
          points="8.660254037844387,4.999999999999999 6.123233995736766e-16,10 -8.660254037844387,4.999999999999999 -8.660254037844389,-4.999999999999997 -1.8369701987210296e-15,-10 8.660254037844384,-5.000000000000004"
        />
      </g>
    </g>
  </g>
</svg>
`;
